File: Vector2D.js

/**
 * @param {number[][]} vector
 */
var Vector2D = function (vector) {

    this.row = 0;
    this.column = 0;
    this.vector = vector;
};

/**
 * @return {number}
 */
Vector2D.prototype.next = function () {
    if (!this.hasNext()) {
        throw `row =  ${this.row} is out of bounds!`;
    }
    return this.vector[this.row][this.column++];
};

/**
 * @return {boolean}
 */
Vector2D.prototype.hasNext = function () {
    this.moveForward();
    return this.row < this.vector.length;
};

Vector2D.prototype.moveForward = function () {
    while (this.row < this.vector.length && this.column === this.vector[this.row].length) {
        this.column = 0;
        this.row++;
    }
};

File: Vector2D.java

public class Vector2D {

    int row;
    int column;
    int[][] vector;

    public Vector2D(int[][] vector) {
        this.vector = vector;
    }

    public int next() {
        if (!hasNext()) {
            throw new IndexOutOfBoundsException("row = " + row + " is out of bounds!");
        }
        return vector[row][column++];
    }

    public boolean hasNext() {
        moveForward();
        return row < vector.length;
    }

    public void moveForward() {
        while (row < vector.length && column == vector[row].length) {
            column = 0;
            row++;
        }
    }
}
